sdk/provisioning/Generator/src/Model/Specification.cs

```diff
@@ -0,0 +1,211 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using Azure.Identity;
+using Azure.ResourceManager;
+
+namespace Azure.Provisioning.Generator.Model;
+
+public abstract partial class Specification : ModelBase
+{
+    /// <summary>
+    /// ArmClient used for talking to the service so it can fetch lists of
+    /// supported versions for resources.
+    /// </summary>
+    private static ArmClient Arm { get; } = new ArmClient(new DefaultAzureCredential());
+
+    public Assembly ArmAssembly { get => ArmType!.Assembly; }
+
+    // Flag indicating we don't need to clean the output directory
+    // because it's merged with another spec that'll handle that for us
+    public bool SkipCleaning { get; protected set; } = false;
+
+    public IList<Resource> Resources { get; private set; } = [];
+
+    public IList<Role> Roles { get; private set; } = [];
+
+    public XmlDocCommentReader DocComments { get; }
+
+    public Dictionary<string, ModelBase> ModelNameMapping { get; } = [];
+    public Dictionary<Type, ModelBase> ModelArmTypeMapping { get; } = [];
+
+    public Specification(string name, Type armEntryPoint)
+        : base(
+            name: name,
+            ns: $"Azure.Provisioning.{name}",
+            armType: armEntryPoint)
+    {
+        Spec = this;
+        DocComments = new XmlDocCommentReader(armEntryPoint.Assembly);
+        TypeRegistry.Register(this);
+    }
+
+    public override string ToString() => $"<Specification {Name}>";
+
+    public void Build()
+    {
+        Analyze();
+        Customize();
+        Lint();
+        ContextualException.WithContext(
+            $"Generating all types for {Namespace}",
+            () =>
+            {
+                // Delete any existing generated code
+                string? path = GetGenerationPath();
+                if (!SkipCleaning && path is not null && Directory.Exists(path)) { Directory.Delete(path, recursive: true); }
+
+                // Regenerate each type
+                HashSet<string> segments = new(Namespace!.Split('.'));
+                foreach (ModelBase? type in ModelNameMapping.Values.Where(p => p is TypeModel || p is EnumModel))
+                {
+                    // Validate that we don't have type/namespace conflicts
+                    if (segments.Contains(type.Name))
+                    {
+                        throw new InvalidOperationException($"Namespace {Namespace} conflicts with type {type.Name} / {type.ArmType!.FullName}");
+                    }
+
+                    type.Generate();
```

Comment:
It's basically an Analyze (via reflection), Customize (by hand), Lint (at least for new libraries) and generate loop.